protocols/mp-ecdsa/src/protocols/sign.rs

```diff
@@ -0,0 +1,336 @@
+use itertools::Itertools;
+use multi_party_ecdsa::gg_2020::party_i::verify;
+use multi_party_ecdsa::gg_2020::state_machine::keygen::{Keygen, LocalKey};
+use multi_party_ecdsa::gg_2020::state_machine::sign::{
+    CompletedOfflineStage, OfflineStage, PartialSignature, SignManual,
+};
+use round_based::async_runtime::watcher::StderrWatcher;
+use sc_client_api::Backend;
+use sp_core::ecdsa::Signature;
+use sp_core::keccak_256;
+use std::collections::HashMap;
+use std::error::Error;
+use std::sync::Arc;
+use tangle_primitives::jobs::{JobId, JobType};
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use tokio::sync::RwLock;
+use webb_gadget::gadget::message::{GadgetProtocolMessage, UserID};
+use webb_gadget::gadget::work_manager::WebbWorkManager;
+use webb_gadget::gadget::{Job, WebbGadgetProtocol, WorkManagerConfig};
+use webb_gadget::protocol::AsyncProtocol;
+use webb_gadget::{Block, BlockImportNotification, FinalityNotification};
+
+pub struct MpEcdsaSigningProtocol<B, BE, KBE, C> {
+    client: MpEcdsaClient<B, BE, KBE, C>,
+    network: GossipNetwork,
+    round_blames: Arc<
+        RwLock<
+            HashMap<
+                <WebbWorkManager as WorkManagerInterface>::TaskID,
+                tokio::sync::watch::Receiver<CurrentRoundBlame>,
+            >,
+        >,
+    >,
+    logger: DebugLogger,
+    account_id: AccountId,
+}
+
+pub async fn create_protocol(
+    _config: &MpEcdsaProtocolConfig,
+) -> Result<MpEcdsaSigningProtocol, Box<dyn Error>> {
+    Ok(MpEcdsaSigningProtocol {})
+}
+
+#[async_trait]
+impl<B: Block, BE: Backend<B>, C: ClientWithApi<B, BE>, KBE: KeystoreBackend> WebbGadgetProtocol<B>
+    for MpEcdsaSigningProtocol<B, BE, KBE, C>
+{
+    async fn get_next_jobs(
+        &self,
+        _notification: &FinalityNotification<B>,
+        now: u64,
+        _job_manager: &ProtocolWorkManager<WebbWorkManager>,
+    ) -> Result<Option<Vec<Job>>, webb_gadget::Error> {
+        let jobs = self
+            .client
+            .query_jobs_by_validator(self.account_id)
+            .await?
+            .into_iter()
+            .filter(|r| matches!(r.job_type, JobType::DKGSignature(..)));
+
+        let mut ret = vec![];
+
+        for job in jobs {
+            let participants = job.participants.expect("Should exist for stage 2 signing");
+            if participants.contains(&self.account_id) {
+                let task_id = job.job_id.to_be_bytes();
+                let task_id = keccak_256(&task_id);
+                let session_id = 0; // We are not interested in sessions for the ECDSA protocol
+                let retry_id = 0; // TODO: query the job manager for the retry id
+
+                let key = job.key.expect("Should exist for stage 2 signing");
+                // TODO for keygen: use bincode2 for serialization
+                let deserialized_key: LocalKey<Secp256k1> =
+                    bincode2::deserialize(&key).expect("Should deserialize");
```

Comment:
This `key` here in the job.key is not the `LocalKey`, Thomas. It is the Public Key of the DKG. The "LocalKey" is, by definition, Local to you, the current validator and never shared, please treat that as a private material. It should only exist in the validator database, encrypted.